core/src/main/scala/scalaz/Cord.scala

```diff
@@ -1,163 +1,55 @@
-  def toStream: Stream[Char] = toVector.toStream
-  def toVector: Vector[Char] = self.foldMap(_.toVector)
+sealed abstract class Cord {
   override def toString: String = {
-    val sb = new java.lang.StringBuilder(self.measure)
-    self.foreach { x =>
-      val _ = sb append x
-    }
+    val sb = new StringBuilder
+    appendTo(sb)
     sb.toString
   }
 
-  /** Transforms each character to a `Cord` according to the given function and concatenates them all into one `Cord`. */
-  def flatMap(f: Char => Cord): Cord = toVector.foldLeft(Cord())((as, a) => as ++ f(a))
+  // FIXME: stack safety. Try to at least make traversal on the right stack safe
+  // (split into two?)
+  private def appendTo(sb: StringBuilder): Unit = this match {
+    case Cord.Branch(a, b) =>
+      a.appendTo(sb): Unit
+      b.appendTo(sb): Unit
+    case Cord.Leaf(s) =>
+      val _ = sb.append(s)
+  }
 
-  /** Returns whether this `Cord` will expand to an empty string. */
-  def isEmpty: Boolean = !nonEmpty
+  // FIXME: used with the implicit... it should be removed
+  @deprecated("use the z / cord interpolators", "7.3")
+  def +:(o: Cord): Cord = Cord.Branch(o, this)
 
-  /** Returns whether this `Cord` will expand to a non-empty string. */
-  def nonEmpty: Boolean = self.iterator.exists(_.nonEmpty)
+  /** Strict evaluation variant of Monoid.append */
+  def ++(o: Cord): Cord = Cord.Branch(this, o)
 }
-
 object Cord {
-  private def cord[A](v: FingerTree[Int, String]): Cord = new Cord(v)
+  private final case class Branch(a: Cord, b: Cord) extends Cord
+  private final case class Leaf(s: String)            extends Cord
 
-  implicit def stringToCord(s: String): Cord = cord(FingerTree.single[Int, String](s))
+  // we could cache all N character strings here with Memo... needs performance
+  // numbers to justify it.
+  def apply(s: String): Cord = if (s.isEmpty) empty else Leaf(s)
 
-  lazy val empty: Cord = apply()
+  val empty: Cord = Leaf("")
 
-  def apply(as: Cord*): Cord = as.foldLeft(cord(FingerTree.empty))(_ ++ _)
-
-  def fromStrings[A](as: Seq[String]): Cord = cord(as.foldLeft(FingerTree.empty[Int, String])((x, y) => x :+ y))
-
-  implicit val sizer: Reducer[String, Int] = UnitReducer((a: String) => a.length)
-
-  def mkCord(sep: Cord, as: Cord*): Cord =
-    if (!as.isEmpty)
-      as.tail.foldLeft(as.head)(_ ++ sep ++ _)
-    else
-      Cord()
-
-  implicit val CordShow: Show[Cord] = new Show[Cord] {
-    override def show(x: Cord) = x
-    override def shows(x: Cord) = x.toString
+  implicit val monoid: Monoid[Cord] = new Monoid[Cord] {
+    def zero: Cord                         = empty
+    def append(f1: Cord, f2: =>Cord): Cord = Branch(f1, f2)
   }
 
-  implicit val CordMonoid: Monoid[Cord] = new Monoid[Cord] {
-    def zero = empty
-    def append(x: Cord, y: => Cord) = x ++ y
-  }
+  implicit val show: Show[Cord] = Show.show(identity)
+  implicit val equal: Equal[Cord] = Equal.equal((a, b) => a.toString == b.toString)
```

Comment:
this _seems_ like it ought to have a more efficient implementation, but later

Reply:
yeah, the problem boils down to whether we want equality to be based on content or structure.